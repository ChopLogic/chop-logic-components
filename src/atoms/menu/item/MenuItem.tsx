import React, { PropsWithChildren, useState } from 'react';
import { ChopLogicOrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import styles from './MenuItem.module.scss';
import { MenuLeaf } from '@atoms/menu/leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import { getClassName } from '@utils';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: ChopLogicOrientationMode;
};

export const MenuItem: React.FC<Props> = ({ item, mode }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([
    styles.subMenuBar,
    { [styles.subMenuBar__horizontal]: mode === ChopLogicOrientationMode.Horizontal },
  ]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  return (
    <SubMenu item={item} isSubMenuOpened={isSubMenuOpened} toggleSubMenu={toggleSubMenu} closeSubMenu={closeSubMenu} mode={mode}>
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} />)}
      </ul>
    </SubMenu>
  );
};
