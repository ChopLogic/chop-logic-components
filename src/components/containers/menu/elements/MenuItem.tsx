import React, { PropsWithChildren } from 'react';

import { ChopLogicMenuItem } from '@/models';

type MenuItemProps = PropsWithChildren & {
  item: ChopLogicMenuItem;
  allItems: ChopLogicMenuItem[];
};

const MenuItem: React.FC<MenuItemProps> = ({ item, allItems }) => {
  const dependentItems = allItems.filter((child) => child.parentId === item.id && child.level - item.level === 1);
  const hasNoChildren = dependentItems.length === 0;

  if (hasNoChildren) {
    return <li role='menuitem'>{item.label}</li>;
  }

  return (
    <li role='menuitem'>
      {item.label}
      <ul role='menuitem'>
        {dependentItems.map((child) => (
          <MenuItem item={child} allItems={allItems} key={child.id} />
        ))}
      </ul>
    </li>
  );
};

export default MenuItem;
