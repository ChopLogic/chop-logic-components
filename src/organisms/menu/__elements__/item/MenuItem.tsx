import { ChopLogicOrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { getClassName } from '@utils';
import { FC, PropsWithChildren, useState } from 'react';

import { MenuLeaf } from '../leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuItem.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: ChopLogicOrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuItem: FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([
    styles.subMenuBar,
    { [styles.subMenuBar__horizontal]: mode === ChopLogicOrientationMode.Horizontal },
  ]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
