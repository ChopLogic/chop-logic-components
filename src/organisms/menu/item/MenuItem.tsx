import { OrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { getClassName } from '@utils';
import { FC, PropsWithChildren, useState } from 'react';

import { MenuLeaf } from '../leaf/MenuLeaf';
import { SubMenu } from '../sub-menu/SubMenu';
import styles from './MenuItem.module.scss';

type Props = PropsWithChildren & {
  item: ChopLogicMenuItem;
  mode: OrientationMode;
  openedOn?: 'hover' | 'click';
};

export const MenuItem: FC<Props> = ({ item, mode, openedOn }) => {
  const isLeaf = !item?.nestedItems?.length;
  const [isSubMenuOpened, setIsSubMenuOpened] = useState(false);
  const subMenuBarClass = getClassName([styles.subMenuBar, { [styles.subMenuBar__horizontal]: mode === OrientationMode.Horizontal }]);

  if (isLeaf) {
    return <MenuLeaf item={item} />;
  }

  const toggleSubMenu = () => {
    setIsSubMenuOpened(!isSubMenuOpened);
  };

  const closeSubMenu = () => {
    setIsSubMenuOpened(false);
  };

  const openSubMenu = () => {
    setIsSubMenuOpened(true);
  };

  return (
    <SubMenu
      item={item}
      isSubMenuOpened={isSubMenuOpened}
      toggleSubMenu={toggleSubMenu}
      closeSubMenu={closeSubMenu}
      openSubMenu={openSubMenu}
      mode={mode}
      openedOn={openedOn}
    >
      <ul className={subMenuBarClass} role='menu' aria-label={item.label}>
        {item?.nestedItems?.map((child) => <MenuItem item={child} key={child.id} mode={mode} openedOn={openedOn} />)}
      </ul>
    </SubMenu>
  );
};
