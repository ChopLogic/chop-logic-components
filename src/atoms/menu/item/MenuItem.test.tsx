import { PropsWithChildren } from 'react';
import { ChopLogicOrientationMode } from '@enums';
import { ChopLogicMenuItem } from '@models';
import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { MenuItem } from './MenuItem';

vi.mock('../sub-menu/SubMenu', () => ({
  SubMenu: ({ children, toggleSubMenu }: PropsWithChildren & { toggleSubMenu: () => void }) => (
    <div data-testid='submenu'>
      <button onClick={toggleSubMenu}>Toggle SubMenu</button>
      {children}
    </div>
  ),
}));

vi.mock('../leaf/MenuLeaf', () => ({
  MenuLeaf: ({ item }: { item: ChopLogicMenuItem }) => <div data-testid='menu-leaf'>{item.label}</div>,
}));

describe('MenuItem', () => {
  const leafItem = {
    id: '1',
    label: 'Leaf Item',
  };

  const nestedItem = {
    id: '2',
    label: 'Nested Item',
    nestedItems: [
      { id: '3', label: 'Child Item 1' },
      { id: '4', label: 'Child Item 2' },
    ],
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<MenuItem item={nestedItem} mode={ChopLogicOrientationMode.Horizontal} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders MenuLeaf for leaf nodes', () => {
    render(<MenuItem item={leafItem} mode={ChopLogicOrientationMode.Horizontal} />);

    const menuLeaf = screen.getByTestId('menu-leaf');
    expect(menuLeaf).toBeInTheDocument();
    expect(menuLeaf).toHaveTextContent('Leaf Item');
  });

  it('renders SubMenu for non-leaf nodes', () => {
    render(<MenuItem item={nestedItem} mode={ChopLogicOrientationMode.Horizontal} />);

    const subMenu = screen.getByTestId('submenu');
    expect(subMenu).toBeInTheDocument();
  });

  it('toggles SubMenu state when toggleSubMenu is called', () => {
    render(<MenuItem item={nestedItem} mode={ChopLogicOrientationMode.Horizontal} />);

    const toggleButton = screen.getByText(/toggle submenu/i);
    fireEvent.click(toggleButton);

    // Check if submenu state changes (mocked submenu renders children conditionally)
    expect(screen.getByTestId('submenu')).toBeInTheDocument();
  });

  it('renders nested items recursively', () => {
    render(<MenuItem item={nestedItem} mode={ChopLogicOrientationMode.Horizontal} />);

    expect(screen.getByText('Child Item 1')).toBeInTheDocument();
    expect(screen.getByText('Child Item 2')).toBeInTheDocument();
  });

  it('passes correct props to SubMenu', () => {
    render(<MenuItem item={nestedItem} mode={ChopLogicOrientationMode.Horizontal} />);

    const subMenu = screen.getByTestId('submenu');
    expect(subMenu).toBeInTheDocument();
    expect(screen.getByText(nestedItem.nestedItems[0].label)).toBeInTheDocument();
    expect(screen.getByText(nestedItem.nestedItems[1].label)).toBeInTheDocument();
  });

  it('passes correct props to MenuLeaf', () => {
    render(<MenuItem item={leafItem} mode={ChopLogicOrientationMode.Horizontal} />);

    const menuLeaf = screen.getByTestId('menu-leaf');
    expect(menuLeaf).toHaveTextContent('Leaf Item');
  });
});
