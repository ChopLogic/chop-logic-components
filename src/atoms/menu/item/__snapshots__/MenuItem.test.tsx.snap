// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuItem > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="submenu"
  >
    <button>
      Toggle SubMenu
    </button>
    <ul
      aria-label="Nested Item"
      class="_subMenuBar_eb1c36 _subMenuBar__horizontal_eb1c36"
      role="menu"
    >
      <div
        data-testid="menu-leaf"
      >
        Child Item 1
      </div>
      <div
        data-testid="menu-leaf"
      >
        Child Item 2
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
